File: CustomApplications/VB.Net/GraphicsHowTo/Primitives/MarkerBatch/MarkerBatchCodeSnippet.vb
#Region "UsingDirectives"
Imports System.Windows.Forms
Imports System.Drawing
Imports System.Collections.Generic
Imports System.IO
Imports AGI.STKGraphics
Imports AGI.STKObjects
Imports AGI.STKUtil
#End Region

Namespace Primitives.MarkerBatch
	Class MarkerBatchCodeSnippet
		Inherits CodeSnippet
		Public Sub New()
            MyBase.New("Primitives\MarkerBatch\MarkerBatchCodeSnippet.vb")
        End Sub

        Public Overrides Sub Execute(ByVal scene As IAgStkGraphicsScene, ByVal root As AgStkObjectRoot)
            Dim markerImageFile As String = New AGI.DataPath(AGI.DataPathRoot.Relative, "Markers/facility.png").FullPath
            ExecuteSnippet(scene, root, markerImageFile)
        End Sub

		' Name        
		' Description 
		' Category    
		' References  
		' Namespaces  
		' EID         
        <AGI.CodeSnippets.CodeSnippet( _
            "DrawASetOfMarkers", _
            "Draw a set of markers", _
            "Graphics | Primitives | Marker Batch Primitive", _
            "System", _
            "System", _
            "AgSTKGraphicsLib~IAgStkGraphicsMarkerBatchPrimitive" _
            )> _
        Public Sub ExecuteSnippet(<AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")> ByVal scene As IAgStkGraphicsScene, <AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")> ByVal root As AgStkObjectRoot, <AGI.CodeSnippets.CodeSnippet.Parameter("markerImageFile", "The image file to use for the markers")> ByVal markerImageFile As String)
            '#Region "CodeSnippet"
            Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager

            Dim positions As Array = New Object() { _
                attachID("$lat1$The latitude of the first marker$", 39.88), attachID("$lon1$The longitude of the first marker$", -75.25), attachID("$alt1$The altitude of the first marker$", 0), _
                 attachID("$lat2$The latitude of the second marker$", 38.85), attachID("$lon2$The longitude of the second marker$", -77.04), attachID("$alt2$The altitude of the second marker$", 0), _
                 attachID("$lat3$The latitude of the third marker$", 29.98), attachID("$lon3$The longitude of the third marker$", -90.25), attachID("$alt3$The altitude of the third marker$", 0), _
                 attachID("$lat4$The latitude of the fourth marker$", 37.37), attachID("$lon4$The longitude of the fourth marker$", -121.92), attachID("$alt4$The altitude of the fourth marker$", 0)}

            Dim markerBatch As IAgStkGraphicsMarkerBatchPrimitive = manager.Initializers.MarkerBatchPrimitive.Initialize()
            markerBatch.Texture = manager.Textures.LoadFromStringUri( _
                markerImageFile)
            markerBatch.SetCartographic(attachID("$planetName$The planet to place the markers on$", "Earth"), positions)

            manager.Primitives.Add(DirectCast(markerBatch, IAgStkGraphicsPrimitive))
            '#End Region

            m_MarkerBatch = DirectCast(markerBatch, IAgStkGraphicsPrimitive)
            OverlayHelper.AddTextBox("A collection of positions is passed to the MarkerBatchPrimitive to " & vbCrLf & _
                                     "visualize markers for each position.", manager)
        End Sub

		Public Overrides Sub View(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			If m_MarkerBatch IsNot Nothing Then
				ViewHelper.ViewBoundingSphere(scene, root, "Earth", m_MarkerBatch.BoundingSphere)
				scene.Render()
			End If
		End Sub

		Public Overrides Sub Remove(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			If m_MarkerBatch IsNot Nothing Then
				Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
				manager.Primitives.Remove(m_MarkerBatch)
				m_MarkerBatch = Nothing

				OverlayHelper.RemoveTextBox(manager)
				scene.Render()
			End If
		End Sub

		Private m_MarkerBatch As IAgStkGraphicsPrimitive
	End Class
End Namespace

File: CustomApplications/CSharp/GraphicsHowTo/Primitives/MarkerBatch/MarkerBatchCodeSnippet.cs
#region UsingDirectives
using System;
using System.Windows.Forms;
using System.Drawing;
using System.Collections.Generic;
using System.IO;
using AGI.STKGraphics;
using AGI.STKObjects;
using AGI.STKUtil;
#endregion

namespace GraphicsHowTo.Primitives.MarkerBatch
{
    class MarkerBatchCodeSnippet : CodeSnippet
    {
        public MarkerBatchCodeSnippet()
            : base(@"Primitives\MarkerBatch\MarkerBatchCodeSnippet.cs")
        {
        }

        public override void Execute(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            string markerImageFile = new AGI.DataPath(AGI.DataPathRoot.Relative, "Markers/facility.png").FullPath;
            Execute(scene, root, markerImageFile);
        }

        [AGI.CodeSnippets.CodeSnippet(
            /* Name        */ "DrawASetOfMarkers",
            /* Description */ "Draw a set of markers",
            /* Category    */ "Graphics | Primitives | Marker Batch Primitive",
            /* References  */ "System",
            /* Namespaces  */ "System",
            /* EID         */ "AgSTKGraphicsLib~IAgStkGraphicsMarkerBatchPrimitive"
            )]
        public void Execute([AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")] IAgStkGraphicsScene scene, [AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")] AgStkObjectRoot root, [AGI.CodeSnippets.CodeSnippet.Parameter("markerImageFile", "The image file to use for the markers")] string markerImageFile)
        {
#region CodeSnippet
            IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;

            Array positions = new object[]
            {
                /*$lat1$The latitude of the first marker$*/39.88, /*$lon1$The longitude of the first marker$*/-75.25, /*$alt1$The altitude of the first marker$*/0,    // Philadelphia
                /*$lat2$The latitude of the second marker$*/38.85, /*$lon2$The longitude of the second marker$*/-77.04, /*$alt2$The altitude of the second marker$*/0, // Washington, D.C.   
                /*$lat3$The latitude of the third marker$*/29.98, /*$lon3$The longitude of the third marker$*/-90.25, /*$alt3$The altitude of the third marker$*/0, // New Orleans
                /*$lat4$The latitude of the fourth marker$*/37.37, /*$lon4$The longitude of the fourth marker$*/-121.92, /*$alt4$The altitude of the fourth marker$*/0    // San Jose
            };

            IAgStkGraphicsMarkerBatchPrimitive markerBatch = manager.Initializers.MarkerBatchPrimitive.Initialize();
            markerBatch.Texture = manager.Textures.LoadFromStringUri(
                markerImageFile);
            markerBatch.SetCartographic(/*$planetName$The planet on which the markers are placed$*/"Earth", ref positions);

            manager.Primitives.Add((IAgStkGraphicsPrimitive)markerBatch);
#endregion

            m_MarkerBatch = (IAgStkGraphicsPrimitive)markerBatch;
            OverlayHelper.AddTextBox(
@"A collection of positions is passed to the MarkerBatchPrimitive to 
visualize markers for each position.", manager);
        }

        public override void View(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            if (m_MarkerBatch != null)
            {
                ViewHelper.ViewBoundingSphere(scene, root, "Earth", m_MarkerBatch.BoundingSphere);
                scene.Render();
            }
        }

        public override void Remove(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            if (m_MarkerBatch != null)
            {
                IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;
                manager.Primitives.Remove(m_MarkerBatch);
                m_MarkerBatch = null;

                OverlayHelper.RemoveTextBox(manager);
                scene.Render();
            }
        }

        private IAgStkGraphicsPrimitive m_MarkerBatch;
    };
}
